keepers/keeper-core/src/lib.rs

```diff
@@ -298,107 +248,135 @@ async fn parallel_confirm_transactions(
     info!(
         "{} transactions submitted, {} confirmed",
         num_transactions_submitted,
-        num_transactions_submitted - executed_signatures.len()
+        confirmed_signatures.len()
     );
+    confirmed_signatures.clone()
 }
 
-pub async fn parallel_execute_instructions(
-    client: &RpcClient,
-    mut instructions: Vec<Instruction>,
+async fn sign_txs(
+    client: &Arc<RpcClient>,
+    transactions: &[&[Instruction]],
     signer: &Arc<Keypair>,
-    retry_count: u16,
-    confirmation_time: u64,
-) -> Result<(), TransactionExecutionError> {
-    /*
-        Note: Assumes all instructions are equivalent in compute, equivalent in size, and can be executed in any order
+) -> Result<Vec<Transaction>, ClientError> {
+    let blockhash = get_latest_blockhash_with_retry(client).await?;
 
-        1) Submits all instructions in parallel
-        2) Waits a bit for them to confirm
-        3) Checks which ones have confirmed, and keeps the ones that haven't
-        4) Repeats retry_count number of times until all have confirmed
+    let signed_txs = transactions
+        .iter()
+        .map(|instructions| {
+            Transaction::new_signed_with_payer(
+                instructions,
+                Some(&signer.pubkey()),
+                &[signer.as_ref()],
+                blockhash,
+            )
+        })
+        .collect();
 
-        Returns all remaining instructions that haven't executed so application can handle
-    */
+    Ok(signed_txs)
+}
 
-    if instructions.is_empty() {
-        return Ok(());
+pub async fn parallel_execute_transactions(
+    client: &Arc<RpcClient>,
+    transactions: &[&[Instruction]],
+    signer: &Arc<Keypair>,
+    retry_count: u16,
+    confirmation_time: u64,
+) -> Result<Vec<Result<(), SendTransactionError>>, TransactionExecutionError> {
+    let mut results = vec![Err(SendTransactionError::ExceededRetries); transactions.len()];
+    let mut retries = 0;
+
+    if transactions.is_empty() {
+        return Ok(results);
     }
 
-    let instructions_per_tx = calculate_instructions_per_tx(client, &instructions, signer)
+    let mut signed_txs = sign_txs(client, transactions, signer)
         .await
-        .map_err(|e| {
-            TransactionExecutionError::ClientError(e.to_string(), instructions.to_vec())
-        })?;
+        .map_err(|e| TransactionExecutionError::ClientError(e.to_string()))?;
 
-    for _ in 0..retry_count {
-        let mut executed_signatures: HashMap<Signature, usize> = HashMap::new();
-        let instruction_batches: Vec<&[Instruction]> =
-            instructions.chunks(instructions_per_tx).collect();
-        parallel_submit_transactions(
-            client,
-            signer,
-            &instruction_batches,
-            &mut executed_signatures,
-        )
-        .await?;
+    while retries < retry_count {
```

Comment:
imo max retries should prolly be incremented on blockhash increment? or related to some time.

retry_count of 3 may be fine for 100 txs, but won't be fine for 2k txs

Reply:
using a retry count of 10, maybe you were looking at a older commit?

Reply:
added logic to fetch blockhash and increment retries only when it's expired